Add option to redeem tokens via exchange in v1 and v2

NakedFool opened the pull request:
## What does this PR do and why?

Fixes - https://github.com/jbx-protocol/juice-interface/issues/333

The intention of this PR is to give users an alternative way to redeem tokens ( via secondary market ) in redeeming modals.

I've created a shared component so it serves many versions of the project, the only thing that is concerning me is the current naming - `RedeemTokenViaExchange` , wdyt? 

## Screenshots or screen recordings


https://user-images.githubusercontent.com/18723426/190599066-4a07903b-dbaa-472d-92a7-5ce80d4d737b.mov



## Acceptance checklist

- [x] I have evaluated the [Approval Guidelines](https://github.com/jbx-protocol/juice-interface/blob/main/CONTRIBUTING.md#approval-guidelines) for this PR.
- [x] I have tested this PR in [all supported browsers](https://github.com/jbx-protocol/juice-interface/blob/main/CONTRIBUTING.md#supported-browsers).
- [x] I have tested this PR in dark mode and light mode (if applicable).

tomquirk commented:
@johnnyd-eth could you pls help review this PR?

johnnyd-eth commented on src/components/v2v3/V2V3Project/V2V3ManageTokensSection/V2RedeemModal.tsx:
```diff
@@ -259,7 +261,15 @@ export default function V2RedeemModal({
                 disabled={totalBalance?.eq(0)}
                 onChange={val => setRedeemAmount(val)}
               />
+              {tokenSymbol && tokenAddress && (
```
**Nit-pick:** We decided a while ago to stick with `{ condition ? <component/> : null }` in cases like this, rather than just `&&`. Sometimes the `&&` can have unintended consequences such as displaying `0` in the UI

NakedFool replied:
Sure! Didn't know there are consequences with approach below. nice learn, thanks!

johnnyd-eth commented on src/components/Project/RedeemTokenViaExchange.tsx:
```diff
@@ -0,0 +1,41 @@
+import AMMPrices from 'components/AMMPrices'
+import { ThemeContext } from 'contexts/themeContext'
+
+export const RedeemTokenViaExchange = ({
```
**Suggestion:** Maybe for consistency with `<AMMPrices />`, we just go with `<RedeemAMMPrices />` ?

NakedFool replied:
Definitely one of the hardest things for me are naming conventions 😄 

pretty good suggestion!

johnnyd-eth commented on src/components/Project/RedeemTokenViaExchange.tsx:
```diff
@@ -0,0 +1,41 @@
+            borderBottom: '1px dashed' + colors.stroke.secondary,
+          }}
+        >
+          Redeem {tokenSymbol} on exchange.
```
**Issue:** this should be marked for translation `<Trans>`

NakedFool replied:
nice catch!

johnnyd-eth commented on src/components/Project/RedeemTokenViaExchange.tsx:
```diff
@@ -0,0 +1,41 @@
+
+  return (
+    <span style={style}>
+      <Tooltip
```
**Suggestion:** I think we should include an `or` here like in the `PayInputSubText` 

<img width="180" alt="Screen Shot 2022-09-24 at 1 55 16 pm" src="https://user-images.githubusercontent.com/96150256/192078994-3e5f987e-9305-40f4-a607-318f0bdb687a.png">

johnnyd-eth replied:
(Then lowercase case the 'R' in 'Redeem' below)

NakedFool replied:
affirmative ser!

johnnyd-eth commented:
All functioning well for me @NakedFool, well done ser! 

On top of the few minor comments, I think we should switch the input and output currency in the AMM links in the redeem instance of the links. I haven't looked at the Sushi link, but for Uniswap, we should switch the `inputCurrency` and `outputCurrency` query params so that the form is prefilled for redeem functionality (token -> ETH), rather than buy (ETH -> token). Make sense?

johnnyd-eth reviewed (changes requested):
.

johnnyd-eth commented:
> All functioning well for me @NakedFool, well done ser!
> 
> On top of the few minor comments, I think we should switch the input and output currency in the AMM links in the redeem instance of the links. I haven't looked at the Sushi link, but for Uniswap, we should switch the `inputCurrency` and `outputCurrency` query params so that the form is prefilled for redeem functionality (token -> ETH), rather than buy (ETH -> token). Make sense?

I think the best approach might be:
- Pass a `mode` prop into `AMMPrices` which is typed to `'buy' | 'redeem'`
- Create a new `amm.ts` file inside `lib` which contains a `generateAmmLink` which takes `mode`, `baseLink` and `tokenAddress` as args
- Switch the input and output currencies within this function depending on `mode`

johnnyd-eth commented:
Also @NakedFool , one more comment. Don't think it's related to your changes, but I'm noticing the 'MAX' button on the redeem amount field isn't working: 

https://user-images.githubusercontent.com/96150256/192417537-f5ed50e7-5dfe-4fc6-9f24-0a1892efb2d8.mp4

Could we fix this?

NakedFool commented:
> > All functioning well for me @NakedFool, well done ser!
> > On top of the few minor comments, I think we should switch the input and output currency in the AMM links in the redeem instance of the links. I haven't looked at the Sushi link, but for Uniswap, we should switch the `inputCurrency` and `outputCurrency` query params so that the form is prefilled for redeem functionality (token -> ETH), rather than buy (ETH -> token). Make sense?
> 
> I think the best approach might be:
> 
> * Pass a `mode` prop into `AMMPrices` which is typed to `'buy' | 'redeem'`
> * Create a new `amm.ts` file inside `lib` which contains a `generateAmmLink` which takes `mode`, `baseLink` and `tokenAddress` as args
> * Switch the input and output currencies within this function depending on `mode`

Excellent approach ser, I've applied this workflow in my latest commit 🍻

NakedFool commented:
> Also @NakedFool , one more comment. Don't think it's related to your changes, but I'm noticing the 'MAX' button on the redeem amount field isn't working:
> 
>  Kapture.2022-09-27.at.12.25.54.mp4 
> Could we fix this?

Uh, this is very weird, locally 'MAX' button works for me but on prod it does not work, huh? I'm investigating this.

localhost

https://user-images.githubusercontent.com/18723426/192507677-2ece5c75-8385-45c4-87fa-bc0ac78041b9.mov

johnnyd-eth commented:
> > Also @NakedFool , one more comment. Don't think it's related to your changes, but I'm noticing the 'MAX' button on the redeem amount field isn't working:
> > Kapture.2022-09-27.at.12.25.54.mp4
> > Could we fix this?
> 
> Uh, this is very weird, locally 'MAX' button works for me but on prod it does not work for me, huh? I'm investigating this.
> 
> localhost
> 
>  screen-capture.5.mov

No problem. Let's create another issue for this and solve it in a follow-up :)

johnnyd-eth commented on src/components/AMMPrices/index.tsx:
```diff
@@ -1,11 +1,13 @@
 import TokenAMMPriceRow from './TokenAMMPriceRow'
 
 type Props = {
+  mode: 'buy' | 'redeem'
```
**Praise**

johnnyd-eth commented on src/components/AMMPrices/index.tsx:
```diff
@@ -39,15 +42,23 @@ export default function AMMPrices({
         exchangeName="Sushiswap"
         tokenSymbol={tokenSymbol}
-        exchangeLink={`https://app.sushi.com/swap?inputCurrency=ETH&outputCurrency=${tokenAddress}`}
+        exchangeLink={generateAMMLink(
```
**Nit-pick:** for functions like this with a few args, it's usually nicer from a devs perspective if they take a single object, so we have a very clear idea of what each arg is doing. Like:

```
generateAMMLink({
mode, 
baseLink: 'https://app.sushi.com',
tokenAddress,
)}
```

NakedFool replied:
Gotcha! Will utilize this style in all future PRs ! 💯

johnnyd-eth commented:
Dito as other PR @NakedFool : All functioning perfectly, great work addressing the comments. Just one more minor comment on this one then we'll get it in

NakedFool commented:
> Dito as other PR @NakedFool : All functioning perfectly, great work addressing the comments. Just one more minor comment on this one then we'll get it in

Applied changes ser! ❤️

johnnyd-eth commented on src/components/AMMPrices/index.tsx:
```diff
@@ -39,15 +42,23 @@ export default function AMMPrices({
-        exchangeLink={`https://app.uniswap.org/#/swap?inputCurrency=ETH&outputCurrency=${tokenAddress}`}
+        exchangeLink={generateAMMLink({
+          mode,
+          baseLink: 'https://app.uniswap.org/#',
```
**Praise**

johnnyd-eth commented:
LFG! Merging
